### PR TITLE
lenet on mnist examples

### PR DESCRIPTION
**Pull request by CarloLucibello**
Also showing progress bars, tensorboard logging and the new DataLoader. 

I'm not sure here what the right balance between features and simplicity would be for an entry point script like "lenet + mnist" is supposed to be. Maybe we can have a much more minimalistic "mlp + mnist"

todo:
- [x] add checkpoints
- [x] add comments to the code

### CODE REVIEW

**Comment by CarloLucibello** [comment]
@MikeInnes @oxinabox 
I get a world age error when using a BSON loaded model, any clue?
``` julia
julia> include("lenet_mnist.jl")

julia> train(loadpath="mtest.bson")
[ Info: Dataset MNIST: 60000 train and 10000 test examples
[ Info: Loaded model: 44426 trainable params
[ Info: Start Training
ERROR: MethodError: no method matching (::BSON.__deserialized_types__.var"##451"{Tuple{Int64,Int64,Int64}})(::Array{Float32,4})
The applicable method may be too new: running in world age 27011, while current world is 27013.
Closest candidates are:
  #8(::Any) at /home/carlo/Git/model-zoo/vision/lenet_mnist/lenet_mnist.jl:22 (method too new to be called from this world context.)
Stacktrace:
 [1] applychain(::Tuple{BSON.__deserialized_types__.var"##451"{Tuple{Int64,Int64,Int64}},Conv{2,4,typeof(relu),Array{Float32,4},Array{Float32,1}},MaxPool{2,4},Conv{2,4,typeof(relu),Array{Float32,4},Array{Float32,1}},MaxPool{2,4},BSON.__deserialized_types__.var"##452",Dense{typeof(relu),Array{Float32,2},Array{Float32,1}},Dense{typeof(relu),Array{Float32,2},Array{Float32,1}},Dense{typeof(identity),Array{Float32,2},Array{Float32,1}}}, ::Array{Float32,4}) at /home/carlo/.julia/packages/Flux/CjjeX/src/layers/basic.jl:30
 [2] (::Chain{Tuple{BSON.__deserialized_types__.var"##451"{Tuple{Int64,Int64,Int64}},Conv{2,4,typeof(relu),Array{Float32,4},Array{Float32,1}},MaxPool{2,4},Conv{2,4,typeof(relu),Array{Float32,4},Array{Float32,1}},MaxPool{2,4},BSON.__deserialized_types__.var"##452",Dense{typeof(relu),Array{Float32,2},Array{Float32,1}},Dense{typeof(relu),Array{Float32,2},Array{Float32,1}},Dense{typeof(identity),Array{Float32,2},Array{Float32,1}}}})(::Array{Float32,4}) at /home/carlo/.julia/packages/Flux/CjjeX/src/layers/basic.jl:32
 [3] eval_loss_accuracy(::DataLoader, ::Chain{Tuple{BSON.__deserialized_types__.var"##451"{Tuple{Int64,Int64,Int64}},Conv{2,4,typeof(relu),Array{Float32,4},Array{Float32,1}},MaxPool{2,4},Conv{2,4,typeof(relu),Array{Float32,4},Array{Float32,1}},MaxPool{2,4},BSON.__deserialized_types__.var"##452",Dense{typeof(relu),Array{Float32,2},Array{Float32,1}},Dense{typeof(relu),Array{Float32,2},Array{Float32,1}},Dense{typeof(identity),Array{Float32,2},Array{Float32,1}}}}, ::Function) at /home/carlo/Git/model-zoo/vision/lenet_mnist/lenet_mnist.jl:58
 [4] (::var"#report#20"{Args,typeof(gpu),DataLoader,DataLoader})(::Int64) at /home/carlo/Git/model-zoo/vision/lenet_mnist/lenet_mnist.jl:124
 [5] #train#17(::Base.Iterators.Pairs{Symbol,String,Tuple{Symbol},NamedTuple{(:loadpath,),Tuple{String}}}, ::typeof(train)) at /home/carlo/Git/model-zoo/vision/lenet_mnist/lenet_mnist.jl:139
 [6] (::var"#kw##train")(::NamedTuple{(:loadpath,),Tuple{String}}, ::typeof(train)) at ./none:0
 [7] top-level scope at REPL[24]:1
```
When loading and using in REPL instead I don't see the issue
```julia
julia> include("lenet_mnist.jl")
[ Info: CUDAdrv.jl failed to initialize, GPU functionality unavailable (set JULIA_CUDA_SILENT or JULIA_CUDA_VERBOSE to silence or expand this message)
train (generic function with 1 method)

julia> mtest = BSON.load("mtest.bson")[:model]
Chain(#8, Conv((5, 5), 1=>6, relu), MaxPool((2, 2), pad = (0, 0, 0, 0), stride = (2, 2)), Conv((5, 5), 6=>16, relu), MaxPool((2, 2), pad = (0, 0, 0, 0), stride = (2, 2)), #9, Dense(256, 120, relu), Dense(120, 84, relu), Dense(84, 10))

julia> dtrain, dtest = get_data(Args());

julia> eval_loss_accuracy(dtest, mtest, cpu)
(loss = 2.2903f0, acc = 8.31)
```

**Comment by oxinabox** [comment]
I have never seen this

**Comment by CarloLucibello** [comment]
I reduced the issue to
```
julia> import BSON

julia> function train()
           model = x -> x
           BSON.@save "model.bson" model    
           model = BSON.load("model.bson")[:model]
           ŷ = model(1)
       end
train (generic function with 1 method)

julia> train()
ERROR: MethodError: no method matching (::BSON.__deserialized_types__.var"#94#95")(::Int64)
The applicable method may be too new: running in world age 27203, while current world is 27204.
Closest candidates are:
  #94(::Any) at REPL[73]:2 (method too new to be called from this world context.)
Stacktrace:
 [1] train() at ./REPL[73]:5
 [2] top-level scope at REPL[74]:1
```
ref. https://github.com/JuliaIO/BSON.jl/issues/69

**Comment by CarloLucibello** [comment]
I think this is ready to go. @dhairyagandhi96 @matsueushi it would be nice to have some feedback from you. On average, I expect the other scripts in this repo to be slimmer than this, but I wanted this particular one to showcase patterns and packages in the ecosystem

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+    "julia.environmentPath": "/home/carlo/Git/model-zoo/vision/lenet_mnist"
+}
```
Comment (matsueushi) [inline .vscode/settings.json]: I don't think settings.json for vscode should be included to this repository.

File: vision/lenet_mnist/lenet_mnist.jl [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        ntot += size(x)[end]
+    end
+    return (loss = l/ntot |> round4, acc = acc/ntot*100 |> round4)
+end
+
+## utility functions
+
+nobs(loader) = sum(size(x)[end] for (x, y) in loader) # == size(loader.data[1])[end] but more generic
+
+num_params(model) = sum(length(p) for p in Flux.params(model))
```
Comment (matsueushi) [inline vision/lenet_mnist/lenet_mnist.jl]:
```suggestion
num_params(model) = sum(length.(Flux.params(model)))
```
It might be my preference...

**Pull request merged**